option/option.go

```diff
@@ -344,6 +345,19 @@ func WithCredentials(creds *google.Credentials) ClientOption {
 	return (*withCreds)(creds)
 }
 
+// WithTokenProvider returns a ClientOption that specifies an
+// [cloud.google.com/go/auth.TokenProvider] to be used as the basis for
+// authentication.
+func WithTokenProvider(tp auth.TokenProvider) ClientOption {
```

Comment:
We already have a universe domain option. I think what we need to make sure is the combination of them resolve as we think they should.

Reply:
This is totally different from `option.WithUniverseDomain`. Every auth flow ALSO needs to accept universe domain, for comparison against `option.WithUniverseDomain`.

But the broader issue here is to make this future-proof. Unlike a struct, an interface is not extensible, therefore no new inputs can be added to `WithTokenProvider`.

Reply:
An alternative solution would be to add an `option` variadic to this func. That would allow users of the default universe to just pass in their `TokenProvider` impl as they would currently.

Reply:
Ah yes I forgot they don't plumb through. From the skimming of docs I found it seems like for options like these it should be assumed GDU. Let's chat offline tmr on this topic though because maybe this influences what types we want to even return from our auth library in the future. Right now a lot of things return TokenProviders but maybe we should default to the whole credentials object if things like the universe domains are going to be so important to future flows.

We could keep this option and add another options that supports the new Credential type as well. This is actually what we have today for the old auth lib. https://pkg.go.dev/google.golang.org/api/option#WithCredentials

Reply:
> Right now a lot of things return TokenProviders but maybe we should default to the whole credentials object.

Yes, I think this would be wise. My view on returning interfaces in the auth module is that it is very limiting and should be avoided in favor of structs wherever possible.

Reply:
> We could keep this option.

I'll keep my request for changes on this PR until we figure out our plan, but keeping this option is OK as long as we provide a way for users to support: "AL-7 Authentication libraries must add the universe_domain property to the [access] token authentication credentials constructor".

Reply:
That item is for the auth libs in particular -- which ones the new lib is caught up with changes should be compliant.